npm_and_yarn/lib/dependabot/npm_and_yarn/package/package_details_fetcher.rb

```diff
@@ -0,0 +1,325 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "json"
+require "excon"
+require "time"
+require "dependabot/package/package_release"
+require "dependabot/package/package_details"
+require "dependabot/npm_and_yarn/package/registry_finder"
+
+module Dependabot
+  module NpmAndYarn
+    module Package
+      class PackageDetailsFetcher
+        extend T::Sig
+
+        sig do
+          params(
+            dependency: Dependabot::Dependency,
+            dependency_files: T::Array[Dependabot::DependencyFile],
+            credentials: T::Array[Dependabot::Credential]
+          ).void
+        end
+        def initialize(
+          dependency:,
+          dependency_files:,
+          credentials:
+        )
+          @dependency = T.let(dependency, Dependabot::Dependency)
+          @dependency_files = T.let(dependency_files, T::Array[Dependabot::DependencyFile])
+          @credentials = T.let(credentials, T::Array[Dependabot::Credential])
+
+          @npm_details = T.let(nil, T.nilable(T::Hash[String, T.untyped]))
+          @dist_tags = T.let(nil, T.nilable(T::Hash[String, String]))
+          @registry_finder = T.let(nil, T.nilable(Package::RegistryFinder))
+        end
+
+        sig { returns(Dependabot::Dependency) }
+        attr_reader :dependency
+
+        sig { returns(T::Array[Dependabot::Credential]) }
+        attr_reader :credentials
+
+        sig { returns(T::Array[Dependabot::DependencyFile]) }
+        attr_reader :dependency_files
+
+        sig { returns(T.nilable(Dependabot::Package::PackageDetails)) }
+        def fetch
+          package_data = fetch_npm_details
+          unless package_data
+            return Dependabot::Package::PackageDetails.new(
+              dependency: @dependency,
+              releases: [], dist_tags:
+              dist_tags
+            )
+          end
+
+          releases = parse_versions(package_data)
+          Dependabot::Package::PackageDetails.new(
+            dependency: @dependency,
+            releases: releases,
+            dist_tags: dist_tags
+          )
```

Comment:
nit, something like this might be easier for the reader to understand.

```suggestion
          Dependabot::Package::PackageDetails.new(
            dependency: @dependency,
            releases: package_data ? parse_versions(package_data) : [],
            dist_tags: dist_tags
          )
```